lib/core/src/Network/Wai/Middleware/Logging.hs

```diff
@@ -285,22 +289,29 @@ sanitize keys bytes = encode' $ case decode' bytes of
 
 instance HasPrivacyAnnotation HandlerLog where
     getPrivacyAnnotation msg = case msg of
-        LogRequestStart -> Public
-        LogRequest _ -> Public
-        LogRequestBody _ _ -> Confidential
-        LogResponse _ _ -> Public
-        LogResponseBody _ -> Confidential
-        LogRequestFinish -> Public
+        LogRequestStart{} -> Public
+        LogRequest{} -> Public
+        LogRequestBody{} -> Confidential
+        LogResponse{} -> Public
+        LogResponseBody{} -> Confidential
+        LogRequestFinish{} -> Public
 
 instance HasSeverityAnnotation HandlerLog where
     getSeverityAnnotation msg = case msg of
         LogRequestStart -> Debug
         LogRequest _ -> Info
         LogRequestBody _ _ -> Debug
-        LogResponse _ status ->
+        LogResponse t _ status -> max (severityFromRequestTime t) $
```

Comment:
I like the idea of warning the user when values are too wrong.
But it may make the logs confusing if log messages change severity and appear or don't appear based on a time value.

My suggestion is make severity constant w.r.t. the metric and then add a trace transformer on top which monitors for abnormal values and then logs messages separately at a higher level. Should I push an example to the branch?

Reply:
Something like 1ab511e ?

Reply:
I'm not sure I like it though... Seems messier with more msg logs.

Can we not use `max (severityFromRequestTime) $` ?

Reply:
I'll drop this (1ab511e), and merge with `max (severityFromRequestTime) $` in the meantime. Hope you don't mind.

```patch
diff --git a/lib/core/src/Network/Wai/Middleware/Logging.hs b/lib/core/src/Network/Wai/Middleware/Logging.hs
index cf98003eb..613804fa6 100644
--- a/lib/core/src/Network/Wai/Middleware/Logging.hs
+++ b/lib/core/src/Network/Wai/Middleware/Logging.hs
@@ -38,7 +38,7 @@ import Control.Applicative
 import Control.Arrow
     ( second )
 import Control.Tracer
-    ( Tracer, contramap, traceWith )
+    ( Tracer, contramap, contramapM, traceWith )
 import Data.Aeson
     ( FromJSON (..), ToJSON (..), Value (..) )
 import Data.ByteString
@@ -82,7 +82,7 @@ withApiLogger
     -> Middleware
 withApiLogger t0 settings app req0 sendResponse = do
     rid <- nextRequestId settings
-    let t = contramap (ApiLog rid) t0
+    let t = classifyReqTime $ contramap (ApiLog rid) t0
     traceWith t LogRequestStart
     start <- getCurrentTime
     (req, reqBody) <- getRequestBody req0
@@ -109,6 +109,18 @@ withApiLogger t0 settings app req0 sendResponse = do
         traceWith t (LogResponse time req status)
         traceWith t (LogResponseBody body)

+
+    transform :: Monad m => (a -> [a]) -> Tracer m a -> Tracer m a
+    transform multiply tr = flip contramapM tr $ \msg -> do
+        mapM_ (traceWith tr) (multiply msg)
+        return msg
+
+    classifyReqTime = transform f
+      where
+        f (LogResponse t req status) = [LogResponseSlow t req status]
+        f _ = []
+
+
 -- | API logger settings
 data ApiLoggerSettings = ApiLoggerSettings
     { _obfuscateKeys :: Request -> [Text]
@@ -245,6 +257,7 @@ data HandlerLog
     | LogRequestBody [Text] ByteString
     -- ^ Request content, with list of sensitive json keys.
     | LogResponse NominalDiffTime Request (Maybe Status)
+    | LogResponseSlow NominalDiffTime Request (Maybe Status)
     | LogResponseBody ByteString
     | LogRequestFinish
     deriving (Generic, Show)
@@ -258,6 +271,14 @@ instance ToText HandlerLog where
             path = T.decodeUtf8 $ rawPathInfo req
             query = T.decodeUtf8 $ rawQueryString req
         LogRequestBody ks body -> sanitize ks body
+        LogResponseSlow time req status ->
+            mconcat [ "Slow response ", method, " ", path, " ", code, " ", text, " in ", tsec ]
+          where
+            method = T.decodeUtf8 $ requestMethod req
+            path = T.decodeUtf8 $ rawPathInfo req
+            code = maybe "???" (toText . statusCode) status
+            text = maybe "Status Unknown" (T.decodeUtf8 . statusMessage) status
+            tsec = T.pack $ show time
         LogResponse time req status ->
             mconcat [ method, " ", path, " ", code, " ", text, " in ", tsec ]
           where
@@ -293,6 +314,7 @@ instance HasPrivacyAnnotation HandlerLog where
         LogRequest{} -> Public
         LogRequestBody{} -> Confidential
         LogResponse{} -> Public
+        LogResponseSlow{} -> Public
         LogResponseBody{} -> Confidential
         LogRequestFinish{} -> Public

@@ -301,11 +323,12 @@ instance HasSeverityAnnotation HandlerLog where
         LogRequestStart -> Debug
         LogRequest _ -> Info
         LogRequestBody _ _ -> Debug
-        LogResponse t _ status -> max (severityFromRequestTime t) $
+        LogResponse _ _ status ->
             case statusCode <$> status of
                 Just s | s == 503 -> Warning
                 Just s | s >= 500 -> Error
                 _ -> Info
+        LogResponseSlow t _ _ -> severityFromRequestTime t
         LogResponseBody _ -> Debug
         LogRequestFinish -> Debug

```